unblob/handlers/archive/hp/ipkg.py

```diff
@@ -0,0 +1,111 @@
+import io
+from pathlib import Path
+from typing import Iterable, Optional, Tuple
+
+from dissect.cstruct import Instance
+from structlog import get_logger
+
+from unblob.extractor import carve_chunk_to_file
+from unblob.file_utils import Endian, File, InvalidInputFormat, StructParser, snull
+from unblob.models import Chunk, Extractor, HexString, StructHandler, ValidChunk
+
+logger = get_logger()
+
+
+C_DEFINITIONS = r"""
+    typedef struct ipkg_file_entry {
+        char file_name[256];
+        uint64 offset;
+        uint64 size;
+        char unknown[4];
+    } ipkg_file_entry_t;
+
+    typedef struct ipkg_header {
+        char magic[4];
+        uint16 major;
+        uint16 minor;
+        uint32 content_table_offset;
+        uint32 unknown_1;
+        uint32 number_of_files;
+        uint32 unknown_2[2];
+        uint32 null;
+        char file_version[256];
+        char product_name[256];
+        char ipkg_name[256];
+        char signature[256];
+    } ipkg_header_t;
+"""
+
+
+def is_valid_header(header: Instance) -> bool:
+    if header.major != 0x1:
+        return False
+    if header.minor != 0x3:
+        return False
+    try:
+        snull(header.ipkg_name).decode("utf-8")
+        snull(header.file_version).decode("utf-8")
+        snull(header.product_name).decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+class IPKGExtractor(Extractor):
+    def __init__(self, header_struct: str):
+        self.header_struct = header_struct
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            i = 0
+            for i, (name, offset, size) in enumerate(self.parse(file)):
+                file.seek(offset)
+                carve_chunk_to_file(
+                    file=file,
+                    chunk=Chunk(start_offset=offset, end_offset=offset + size),
+                    carve_path=outdir.joinpath(Path(snull(name).decode("utf-8"))),
+                )
+
+    def parse(self, file: File) -> Iterable[Tuple[bytes, int, int]]:
+        header = self._struct_parser.parse("ipkg_header_t", file, Endian.LITTLE)
+        file.seek(header.content_table_offset, io.SEEK_SET)
+
+        entries = [
+            (entry.file_name, entry.offset, entry.size)
+            for entry in (
+                self._struct_parser.parse("ipkg_file_entry_t", file, Endian.LITTLE)
+                for _ in range(header.number_of_files)
+            )
+        ]
+
+        return entries
```

Comment:
Same feeling here, I think this can be simplified by putting everything in `extract` and using a single clear loop.